Add TECO AC

hcoohb opened the pull request:
Hi,

I have added the protocol to send and receive Teco A/C messages.
That would be great if that can be merged to master.
First PR here, so let me know if some things need to be changed.

Regards,
Fabien

hcoohb committed:
Add TECO AC

hcoohb committed:
Modif Makefile

hcoohb committed:
and the other makefile...

crankyoldgit committed:
Update IRutils.cpp

crankyoldgit committed:
Style cleanup for Teco temp handling

crankyoldgit commented:
@hcoohb Wow. Thanks for this. I've got very little changes etc to add to this.
It looks pretty good so far.

I'd like to add some unit tests for this protocol to ensure it does what we think it does.
For that, can you please provide some "raw data" for a few messages and a full description of what you think should be setting in each message etc.
e.g. "Off", "On, Cool 25C Fan High, Swing toggle", ... etc.

We also need to update the examples too to reflect the new additions. I'm happy to do that later etc.

crankyoldgit commented on src/ir_Teco.cpp:
```diff
@@ -0,0 +1,264 @@
+  }
+  // Special modes.
+  if (getSwing()) {
+    result += ", Swing: Toggle";
```
Is it really the case that it toggles the mode, or is it a On/Off flag?
And should we be `return`ing here or just continuing with the decode?
e.g. is changing Swing really incompatible with Sleep, Mode, & Temp?

hcoohb replied:
No, you are right, this is a On/Off flag, not a pure toggle as such.
Nop, my bad, no need to return just yet... we should continue the decode.

hcoohb replied:
The Swing is indeed a On/Off flag..
And you are right, we should continue the decode... my bad

crankyoldgit commented on src/ir_Teco.cpp:
```diff
@@ -0,0 +1,264 @@
+  }
+  if (getSleep()) {
+    result += ", Sleep: Toggle";
+    return result;
```
Ditto previous comment.

hcoohb replied:
Same as above, my bad. On/Off flag, no need to return just yet.

hcoohb replied:
There again. Only on/off flag and decode can continue

crankyoldgit commented on src/ir_Teco.cpp:
```diff
@@ -0,0 +1,264 @@
+    case kTecoAuto:
+    case kTecoCool:
+    case kTecoDry:
+    case kTecoFan:
```
Later on you indicate that `kTecoFan` mode doesn't have a temperature setting. I'm assuming the bits containing the temp are set to something, possibly a special value. If so, we should set that here too.

hcoohb replied:
In Fan mode, the temperature bits do not change. They keep whichever state they were when in cool or heat mode

crankyoldgit commented on src/ir_Teco.cpp:
```diff
@@ -0,0 +1,264 @@
+    default:
+      result += " (UNKNOWN)";
+  }
+  if (getMode() != kTecoFan)  // Fan mode doesn't have a temperature.
```
I'm guessing the remote doesn't display the temp when in `kTecoFan` mode, but it might be wise to show this value, unless it is some special value when in fan mode. Is there a special value for the temp when in Fan mode? (see earlier comment)

hcoohb replied:
No, Fan mode does not change the temperature bits

crankyoldgit committed:
Style cleanup and some Unit tests for Teco protocol.
* Add some unit tests to cover basic functionality of Teco protocol.
* Improve message decoding slightly (Correct expected min size)
* C++ style lint/cleanup/improvements etc.
* Make changes to example code for full Teco support.

Ref #622

hcoohb committed:
Correct display of decode. Change state of reset (swing:off)

hcoohb committed:
Add real message from Raw Data

hcoohb commented:
@crankyoldgit Thank you so much for the very quick and in-depth review!
I was giving it a go to try to write the unit test, but you were much faster than me...
I corrected a few things following your comments and I have added a "real massage from raw data". Is one enough or do we need several?

Thanks again!

crankyoldgit committed:
Fine tune Teco before merge.
* Add a message reconstruction test to verify class can rebuild message.
* Add some comments.
* Minor code cleanup.

Ref #622

crankyoldgit committed:
Update Contrib info for Teco.
Also bump IRMQTTServer version number.

crankyoldgit commented:
> I have added a "real massage from raw data". Is one enough or do we need several?

Yep. One is fine. Feel free to add some more _raw_ ones if you want. I like to have something that confirms we can decode a _real_ message. Hence, one is okay. More is almost always better.

About the only thing I'd add is some more tests that it decodes/creates message codes correctly via the class methods.

e.g. Confirm via the display on your remote == some code which also has a `.toString()` representation that has the same info as your remote.
That can be done via `ac.setRaw(code)` rather than the entire raw data method.

Before I merge this, can you confirm `IRrecvDumpV2` decodes your remote correctly (inc. the text description etc) as a TECO message, and that the sendTeco() method controls your A/C as expected?

hcoohb commented on src/ir_Teco.h:
```diff
@@ -0,0 +1,135 @@
+const uint64_t kTecoTimerTenHr = 0b00000000000000000000110000000000000;
+const uint64_t kTecoTimerOn =    0b00000000000000000001000000000000000;
+const uint64_t kTecoTimerUniHr = 0b00000000000000011110000000000000000;
+const uint64_t kTecoReset =      0b01001010000000000000010000000000000;
```
Is there any particular reason why you changed the Reset state to have the timer at 10 hours? (although the timer is not on).
I would have put it at: 0b01001010000000000000000000000000000

crankyoldgit replied:
Yes, if you look here: https://github.com/markszabo/IRremoteESP8266/pull/622/commits/cd0d60899afb7d261ff0e3bc84653e159c86f9b3#diff-47417b72c15b4b042cc8d367a2a69633R240 in this unit test, I try to construct the message code you added via the `rawData` unit test.
As we don't _yet_ have the timer functions, I couldn't set that bit _yet_ via published/existing methods. I did have a look at implementing them, but I couldn't understand your comments well enough to do so. i.e. the comments describing the time/hour structure.. e.g. it wasn't clear to me.

hcoohb committed:
Add second real world example

hcoohb commented:
I just have tested the receiving from remote and sending to AC  with the newest changes: 
It still works beautifully. Everything works well.
And I confirm that IRrecvDumpV2 recognize the remote IR as a Teco message and display the right information about the signal received.

crankyoldgit committed:
Address a merge conflict.

crankyoldgit commented:
Thanks for the confirmation. I'll merge as soon as the travis check passes after I resolved a merge conflict.

crankyoldgit commented:
FYI, v2.5.6 has just been release which includes the changes/improvements mentioned.
